### PR TITLE
Make topk deterministic under deterministic algorithms

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #186653

When torch.use_deterministic_algorithms(True) is enabled, make topk resolve tied values with stable index tie-breaking.

CPU topk now compares by value and then index in deterministic mode. CUDA keeps the existing top-k selection path and uses stable sorting for the selected top-k outputs so threshold ties preserve input-index order without falling back to a full sort on NVIDIA. ROCm uses stable sort in deterministic mode because its current top-k gather reserves tie slots through atomics.

This gives deterministic mode a topk behavior analogous to scatter_add: fast default kernels remain available, while deterministic mode uses stable algorithms when requested.

Test Plan:

- Rebuilt PyTorch from source with CUDA enabled.

- python -m py_compile test/test_sort_and_select.py torch/__init__.py torch/_torch_docs.py

- python test/test_sort_and_select.py -k topk_deterministic

- python test/test_sort_and_select.py -k topk

- git diff --check HEAD^

- lintrunner --config=.lintrunner.toml --skip=PYREFLY aten/src/ATen/native/TopKImpl.h aten/src/ATen/native/cuda/TensorTopK.cpp test/test_sort_and_select.py torch/__init__.py torch/_torch_docs.py

cc @manuelcandales @angelayi

### CODE REVIEW

**Comment by sanketpurandare** [comment]
TopK perf benchmark details

I ran the same CUDA-event microbenchmark on the parent and on this PR. The benchmark measures `torch.topk(x, k, dim=-1, largest=True, sorted=...)` with 30 warmup iterations and 100 timed iterations.

Note: the benchmark was captured just before the final ghstack hash rewrite. The measured trees correspond to the same parent/current code after ghstack rewrote the commits.

Summary:

- `sorted=False`: deterministic mode is effectively unchanged in this benchmark set. Median current-build overhead was `+0.03%`.
- `sorted=True`: deterministic mode pays the expected stable selected-output sort cost. Median current-build overhead was `+3.24%`; max in these cases was `+8.8%` for small `k=8` router-like shapes.
- Current default mode vs parent default mode did not show a regression in this run. The parent/current numbers are noisy, but current was generally faster in this sample.

Benchmark script:

```python
import torch

assert torch.cuda.is_available()
print(f"version,{torch.__version__},{torch.version.git_version}")
print("case,dtype,pattern,shape,k,sorted,deterministic,ms")

def make(shape, dtype, pattern):
    if pattern == "random":
        torch.manual_seed(123)
        return torch.randn(shape, device="cuda", dtype=dtype)
    if pattern == "ties":
        return torch.ones(shape, device="cuda", dtype=dtype)
    raise AssertionError(pattern)

def bench(x, k, sorted_flag, deterministic, iters=100, warmup=30):
    torch.use_deterministic_algorithms(deterministic)
    for _ in range(warmup):
        torch.topk(x, k, dim=-1, largest=True, sorted=sorted_flag)
    torch.cuda.synchronize()
    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)
    start.record()
    for _ in range(iters):
        torch.topk(x, k, dim=-1, largest=True, sorted=sorted_flag)
    end.record()
    torch.cuda.synchronize()
    return start.elapsed_time(end) / iters

cases = [
    ("router_small", (8192, 256), 8, "random"),
    ("router_large", (32768, 256), 8, "random"),
    ("mb_many_rows", (4096, 1024), 8, "random"),
    ("mb_threshold_random", (2, 20000), 128, "random"),
    ("mb_threshold_ties", (2, 20000), 128, "ties"),
    ("large_selected_random", (2, 10000), 5000, "random"),
    ("large_selected_ties", (2, 10000), 5000, "ties"),
    ("mb_large_selected_ties", (2, 20000), 5000, "ties"),
]

try:
    for dtype in (torch.float32, torch.bfloat16):
        for name, shape, k, pattern in cases:
            x = make(shape, dtype, pattern)
            for sorted_flag in (True, False):
                for deterministic in (False, True):
                    ms = bench(x, k, sorted_flag, deterministic)
                    print(
                        f"{name},{dtype},{pattern},{shape},{k},"
                        f"{sorted_flag},{deterministic},{ms:.6f}"
                    )
finally:
    torch.use_deterministic_algorithms(False)
```

Current PR: deterministic vs non-deterministic mode, same build:

| case | dtype | sorted | non-det ms | det ms | det delta |
|---|---:|---:|---:|---:|---:|
| router_small | float32 | True | 0.108579 | 0.118125 | +8.8% |
| router_small | float32 | False | 0.099021 | 0.099123 | +0.1% |
| router_large | float32 | True | 0.401357 | 0.433250 | +7.9% |
| router_large | float32 | False | 0.376444 | 0.376551 | +0.0% |
| mb_many_rows | float32 | True | 0.149117 | 0.153001 | +2.6% |
| mb_many_rows | float32 | False | 0.141713 | 0.141844 | +0.1% |
| mb_threshold_random | float32 | True | 0.069677 | 0.069638 | -0.1% |
| mb_threshold_random | float32 | False | 0.061834 | 0.061900 | +0.1% |
| mb_threshold_ties | float32 | True | 0.068635 | 0.068539 | -0.1% |
| mb_threshold_ties | float32 | False | 0.061158 | 0.061167 | +0.0% |
| large_selected_random | float32 | True | 0.133212 | 0.137570 | +3.3% |
| large_selected_random | float32 | False | 0.046561 | 0.046510 | -0.1% |
| large_selected_ties | float32 | True | 0.137249 | 0.141561 | +3.1% |
| large_selected_ties | float32 | False | 0.060200 | 0.060210 | +0.0% |
| mb_large_selected_ties | float32 | True | 0.138883 | 0.143342 | +3.2% |
| mb_large_selected_ties | float32 | False | 0.061447 | 0.061566 | +0.2% |
| router_small | bfloat16 | True | 0.106375 | 0.115198 | +8.3% |
| router_small | bfloat16 | False | 0.097390 | 0.097549 | +0.2% |
| router_large | bfloat16 | True | 0.398425 | 0.423512 | +6.3% |
| router_large | bfloat16 | False | 0.371554 | 0.371678 | +0.0% |
| mb_many_rows | bfloat16 | True | 0.103100 | 0.106240 | +3.0% |
| mb_many_rows | bfloat16 | False | 0.095460 | 0.095652 | +0.2% |
| mb_threshold_random | bfloat16 | True | 0.051110 | 0.051128 | +0.0% |
| mb_threshold_random | bfloat16 | False | 0.044936 | 0.045016 | +0.2% |
| mb_threshold_ties | bfloat16 | True | 0.050277 | 0.049968 | -0.6% |
| mb_threshold_ties | bfloat16 | False | 0.045700 | 0.045429 | -0.6% |
| large_selected_random | bfloat16 | True | 0.103699 | 0.107917 | +4.1% |
| large_selected_random | bfloat16 | False | 0.038894 | 0.038883 | -0.0% |
| large_selected_ties | bfloat16 | True | 0.100977 | 0.105220 | +4.2% |
| large_selected_ties | bfloat16 | False | 0.038315 | 0.038268 | -0.1% |
| mb_large_selected_ties | bfloat16 | True | 0.106492 | 0.111162 | +4.4% |
| mb_large_selected_ties | bfloat16 | False | 0.046114 | 0.045442 | -1.5% |

Parent default vs current default:

| case | dtype | sorted | parent ms | current ms | delta |
|---|---:|---:|---:|---:|---:|
| router_small | float32 | True | 0.125955 | 0.108579 | -13.8% |
| router_small | float32 | False | 0.103136 | 0.099021 | -4.0% |
| router_large | float32 | True | 0.453744 | 0.401357 | -11.5% |
| router_large | float32 | False | 0.429144 | 0.376444 | -12.3% |
| mb_many_rows | float32 | True | 0.158976 | 0.149117 | -6.2% |
| mb_many_rows | float32 | False | 0.145781 | 0.141713 | -2.8% |
| mb_threshold_random | float32 | True | 0.081499 | 0.069677 | -14.5% |
| mb_threshold_random | float32 | False | 0.062721 | 0.061834 | -1.4% |
| mb_threshold_ties | float32 | True | 0.070491 | 0.068635 | -2.6% |
| mb_threshold_ties | float32 | False | 0.080370 | 0.061158 | -23.9% |
| large_selected_random | float32 | True | 0.149584 | 0.133212 | -10.9% |
| large_selected_random | float32 | False | 0.049428 | 0.046561 | -5.8% |
| large_selected_ties | float32 | True | 0.153702 | 0.137249 | -10.7% |
| large_selected_ties | float32 | False | 0.060164 | 0.060200 | +0.1% |
| mb_large_selected_ties | float32 | True | 0.150855 | 0.138883 | -7.9% |
| mb_large_selected_ties | float32 | False | 0.069135 | 0.061447 | -11.1% |
| router_small | bfloat16 | True | 0.106422 | 0.106375 | -0.0% |
| router_small | bfloat16 | False | 0.110398 | 0.097390 | -11.8% |
| router_large | bfloat16 | True | 0.460187 | 0.398425 | -13.4% |
| router_large | bfloat16 | False | 0.428820 | 0.371554 | -13.4% |
| mb_many_rows | bfloat16 | True | 0.116715 | 0.103100 | -11.7% |
| mb_many_rows | bfloat16 | False | 0.099891 | 0.095460 | -4.4% |
| mb_threshold_random | bfloat16 | True | 0.059557 | 0.051110 | -14.2% |
| mb_threshold_random | bfloat16 | False | 0.051181 | 0.044936 | -12.2% |
| mb_threshold_ties | bfloat16 | True | 0.052889 | 0.050277 | -4.9% |
| mb_threshold_ties | bfloat16 | False | 0.051166 | 0.045700 | -10.7% |
| large_selected_random | bfloat16 | True | 0.106556 | 0.103699 | -2.7% |
| large_selected_random | bfloat16 | False | 0.042153 | 0.038894 | -7.7% |
| large_selected_ties | bfloat16 | True | 0.113215 | 0.100977 | -10.8% |
| large_selected_ties | bfloat16 | False | 0.038270 | 0.038315 | +0.1% |
| mb_large_selected_ties | bfloat16 | True | 0.120620 | 0.106492 | -11.7% |
| mb_large_selected_ties | bfloat16 | False | 0.050073 | 0.046114 | -7.9% |

File: aten/src/ATen/native/cuda/TensorTopK.cpp [hunk 1]
```diff
@@ -83,9 +85,18 @@ TORCH_IMPL_FUNC(topk_out_cuda)
   checkAllSameGPU(__func__, {topK_arg, indices_arg, input_arg});
 
   dim = at::maybe_wrap_dim(dim, self);
+  // See Note [Enabling Deterministic Operations]
+  const bool deterministic = at::globalContext().deterministicAlgorithms();
+
+#if defined(USE_ROCM)
```
Comment (drisspg) [inline aten/src/ATen/native/cuda/TensorTopK.cpp:91]: why the rocm speical case?

Comment (sanketpurandare) [reply aten/src/ATen/native/cuda/TensorTopK.cpp:91]: rcom uses stable full sort in deterministic mode because its gather path can reserve tie slots through atomics.

File: aten/src/ATen/native/TopKImpl.h [hunk 2]
```diff
@@ -48,42 +65,12 @@ void topk_impl_loop(
       queue[j].second = j;
     }
 
-    // we want nan to be sorted as top for numpy compatibility
```
Comment (drisspg) [inline aten/src/ATen/native/TopKImpl.h:51]: nit maybe leave this coment I think its helpful

File: test/test_sort_and_select.py [hunk 3]
```diff
@@ -514,6 +515,61 @@ def test_topk_quantized_scalar_input(self):
...
+            )
+
+            _, indices = torch.topk(x, 3, dim=-1, largest=True, sorted=False)
+            self.assertEqual(
+                indices.sort(dim=-1).values,
+                torch.tensor([[0, 1, 3], [0, 1, 4]], device=device),
+            )
+
+    @onlyCUDA
+    def test_topk_deterministic_tie_breaking_cuda_multiblock(self, device):
```
Comment (drisspg) [inline test/test_sort_and_select.py:572]: can we confirm stable=True preserves index order for equal keys for 129 <= k <= 4096, or add a tied-value test with k > 128?

File: test/test_sort_and_select.py [hunk 4]
```diff
@@ -514,6 +515,61 @@ def test_topk_quantized_scalar_input(self):
...
+            self.assertEqual(
+                indices,
+                torch.arange(k, device=device).expand_as(indices),
+            )
+
+            _, indices = torch.topk(x, k, dim=-1, largest=False, sorted=True)
+            self.assertEqual(
+                indices,
+                torch.arange(k, device=device).expand_as(indices),
+            )
```
Comment (drisspg) [inline test/test_sort_and_select.py]: also maybe we should add a nan test as part of this

**Review by drisspg — APPROVED** [review]

File: aten/src/ATen/native/cuda/TensorTopK.cpp (hunk 1 above)
Comment (drisspg) [reply aten/src/ATen/native/cuda/TensorTopK.cpp:91]: hmmm so this will likely be a big perf hit?

what used to happen with deterministic mode? we would just silently be non determ right? if so I think thats find and worth the hit

Comment (sanketpurandare) [reply aten/src/ATen/native/cuda/TensorTopK.cpp:91]: Yes with deterministic mode we were being silently non determ

Comment (sanketpurandare) [reply aten/src/ATen/native/cuda/TensorTopK.cpp:91]: I added a comment explaining the rocm case and kept the full stable-sort fallback scoped to rocm deterministic mode only. Default mode is unchanged, and nvidia deterministic mode still uses gather + stable sort of the selected top-k outputs rather than full sorting the input.

File: aten/src/ATen/native/TopKImpl.h (hunk 2 above)
Comment (sanketpurandare) [reply aten/src/ATen/native/TopKImpl.h:51]: Left as is on top

File: test/test_sort_and_select.py (hunk 3 above)
Comment (sanketpurandare) [reply test/test_sort_and_select.py:572]: Added a multiblock deterministic tie test for k covering [128, 129, 4096]. It uses all-equal inputs and checks that both largest/smallest sorted topk return indices arange(k), so it covers stable tie-breaking beyond the small-sort path.
